### PR TITLE
[#36] Add ability for existing users to log in to the app

### LINKED ISSUE #36: Add ability to log in to the app

### PR DESCRIPTION
**Pull request by tiphainetiphaine**
Resolves #36 

- an ~existing user (in the database) can now log in to the app. The "new sign up" flow is there but currently disabled, as this is outside the scope of this issue
- used https://www.c-sharpcorner.com/blogs/building-a-login-flow-with-net-maui as a reference for the login 
- I added tests for the credentials check method, with an in memory mocked database
- also tested the full flow using iOS 17 (iPhone 15) on the simulator
- as mentioned in the code, currently the authentication is currently not persisted. This means users have to log in every time they use the app. I believe this is usually done with an authentication token. We could do it using some kind of local storage, but for iOS this apparently requires a paid developer account hence why this hasn't been done
- since this is for demonstration purposes I think this is enough to close this issue, if we want / need to persist the auth, we can add a separate issue

### CODE REVIEW

File: HaulageApplication/HaulageApp/Models/User.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+using SQLite;
```
Comment (MxmIv) [inline HaulageApplication/HaulageApp/Models/User.cs]: We should not be using SQLite annotations. This project is using  Entity Framework Core (EF Core) with the SQLServer which runs locally. 

as can be seen in the MauiProgram.cs: 
      `  builder.Services.AddDbContext<HaulageDbContext>(options => options.UseSqlServer(connectionString));`

File: HaulageApplication/HaulageApp/ViewModels/LoginViewModel.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        _context = dbContext;
+    }
+    
+    public string Email { get; set; }
+    public string Password { get; set; }
+    
+    [RelayCommand]
+    private async Task Login()
+    {
+        if (IsCredentialCorrect(Email, Password))
```
Comment (MxmIv) [inline HaulageApplication/HaulageApp/ViewModels/LoginViewModel.cs]: Add try-catch block to catch database connection errors

Comment (MxmIv) [reply HaulageApplication/HaulageApp/ViewModels/LoginViewModel.cs]: Also, should we be handling cases if Email and Password are `null`. We should throw an error there?

File: HaulageApplication/HaulageAppTests/CredentialsTest.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        
+        CreateContext(options);
+        
+        using (var context = new HaulageDbContext(options))
+        {
+            var viewmodel = new LoginViewModel(context);
+            Assert.False(viewmodel.IsCredentialCorrect("who", "1234"));
+        }
+    }
+}
```
Comment (MxmIv) [inline HaulageApplication/HaulageAppTests/CredentialsTest.cs:57]: Add a text to handle null inputs

**Review by MxmIv — CHANGES_REQUESTED** [review]

File: HaulageApplication/HaulageApp/ViewModels/LoginViewModel.cs (hunk 2 above)
Comment (tiphainetiphaine) [reply HaulageApplication/HaulageApp/ViewModels/LoginViewModel.cs]: I think a better way to handle nulls is to set any blank inputs to "", then you don't need to actually handle them. IT will be handled as an incorrect attempt.

File: HaulageApplication/HaulageAppTests/CredentialsTest.cs (hunk 3 above)
Comment (tiphainetiphaine) [reply HaulageApplication/HaulageAppTests/CredentialsTest.cs:57]: following my previous changes (adding default values of "" for both fields), those properties can't be null.

**Comment by tiphainetiphaine** [comment]
Screenshot 2024-08-16 at 14 18 28
Screenshot 2024-08-16 at 14 24 27

Case "internet connection" error - I stopped the docker container to test it.
The other case is update for empty fields (now with default value).
I think this is better than throwing exceptions.

**Review by FMulligan — APPROVED** [review]

File: HaulageApplication/HaulageAppTests/CredentialsTest.cs (hunk 3 above)
Comment (MxmIv) [reply HaulageApplication/HaulageAppTests/CredentialsTest.cs:57]: we should still add the test and throw an exception of we will get null.

Comment (MxmIv) [reply HaulageApplication/HaulageAppTests/CredentialsTest.cs:57]: yeah , if we initialise them to empty stings we are ok not not having that

**Review by MxmIv — APPROVED** [review]
Thanks for including screenshots! LGTM
